### ISSUE TITLE
Sine Oscillator Aliasing with LFO on Pitch

### ISSUE THREAD

**Issue by unfa** [surge-synthesizer/surge#3784]
**Is your feature request related to a problem? Please describe.**
Sometimes, when using osciallators at very high-frequencies (>10kHz) I can hear aliasing occuring, even on a sine wave. This is especially apparent when modulating the pitch.

**Describe the solution you'd like:**
I think that maybe adding an option to oversample the oscillators (to avoid aliasing) and then lowpassing them before feeding the signal down the line could solve this issue. 

**Describe alternatives you've considered:**
Maybe including an oversampling option for the entire signal chain?

**Comment by mkruselj** [comment]
Oscillators are already 2x oversampled. This is a hard constant in the codebase, and if it's changed some oscillators don't sound correctly anymore (for example the Classic oscillator gets completely borked).

**Comment by unfa** [comment]
Thank you for the explanation, that sounds like great care has been taken to make Surge sound as clean as possible without being too heavy.

I am using the Sine Oscillator (not a wavetable).

When playing a sine wave with LFO modulating the pitch that'd oscillate around 10 kHz touching or exceeding 20 kHz, I can hear some additional tones that sound like aliasing in the range above 12 kHz. I'm not sure I can measure this, the spectrogram readouts are not clean, but I am not sure if it's the nature of the tool, or actual problems being shown. I cannot hear or measure that on static tones. I wonder if this could be some problem with LFO interpolation, that only shows itself at these frequencies?

**Comment by baconpaul** [comment]
Can you attach a patch (just store, zip the fxp, attach here) and we can look. We might be doing something dumb!

**Comment by baconpaul** [comment]
And yeah lfo is linear over the block so that may alias depenging on lfo rate. If you want fm don’t use the LFOs. But I would like to see the patch still!

**Comment by unfa** [comment]
Sure, I'll prepare a test patch!

**Comment by unfa** [comment]
Saturation makes the aliasing much more audible, if you increase the LFO rate to 0.5 Hz and disable the Waveshaper, you should be able to hear it too.

[Aliasing Test.fxp.zip](https://github.com/surge-synthesizer/surge/files/5904879/Aliasing.Test.fxp.zip)

**Comment by baconpaul** [comment]
Thanks!

**Comment by mkruselj** [comment]
So the sine oscillator is only 2x oversampled, but it is NOT band-limited (in the way Classic, Wavetable and S&H Noise are, which are BLIT oscillators). This is kinda by design, and we don't have anything in the cards regarding bumping up oversampling rate or additional band-limiting (which would in fact change the sound of sine oscillator in certain ranges).

I am moving this to "currently unscheduled" milestone.

**Comment by baconpaul** [comment]
There was actually one thing I wanted to check here @mkruselj so moving it back into 1.9.0 so I remember.

**Comment by baconpaul** [comment]
OK the one thing I wanted to check wasn't it

Yeah I agree this one is hard to anti-alias up that high with the way sine works.

But interestingly I just tried it with the new Modern oscillator in Sine configuration and the aliasing drops a lot. Which makes sense, since that uses a totally different algo which is meant to combat aliasing. 

That algo, alas, doesn't work well with the wave shaping and the like we do in sine. but something to consider. Not for 1.9 though so I agree I'm going to bump this into a later milestone.

**Comment by baconpaul** [comment]
by the way: an excellent patch. You can hear the mirror frequencies clear as day in the sine oscillator version. Thanks as always for the great bug reports.

**Comment by mkruselj** [comment]
So I think we should close this issue. We are not going to change the Sine osc, and Modern osc has a sine that is cleaner up high (which is what unfa was trying to do - just have a modulated sine, and not necessarily one of 20-something other waveforms). The potential solution for all the other Sine osc waveforms would be custom waveshaper curves that can reshape that sine into any of those other Sine osc shapes (#1964).

**Comment by baconpaul** [comment]
I’m not so sure - I would like to leave this open to see if I can change the core sin/cos evaluator from the approx we used to a less alias-y one. So let’s leave it for now

**Comment by mkruselj** [comment]
@baconpaul Claude said that the Padé approximations used are already very clean (-100...-127 dBFS), the aliases are coming from the fact that `omega` value is being held per block instead of interpolated (which is the reason behind Sine in Modern sounding cleaner, phase increment is ramped linearly across block there). And of course, unfa using a waveshaper bumped about 20 dB in that test patch exacerbated the issue, especially since that waveshaper is set to Soft mode, which is a LUT at 2xOS, so those aliases are inherent to that design - we would need to ADAA instead of LUT to get it cleaner. Definitely not a 1.x change. For XT2 we should ADAA all the waveshapers and never LUT them.

With waveshaper disabled in that test patch, and lerping across block implemented, aliasing improves from -60 dBc to about -100 dBc with that same test patch. We should at least take that win for 1.4, wdyt? (Not to mention lerping the `omega` brings it in line with phase accumulator lerping across block in other oscillator types.)

<img width="1920" height="1160" alt="Image" src="https://github.com/user-attachments/assets/e44720ee-c6e6-45a7-8747-27593b70fffa" />
<img width="1920" height="1160" alt="Image" src="https://github.com/user-attachments/assets/54d78d85-95cf-4099-bcac-73a8b7dc7028" />

**Comment by mkruselj** [comment]
Also seems like FM2 and FM3 have the same issue where `omega` is not lerped across the block. This would require an update in `sst-basic-blocks` for the `SurgeQuadrOscillator`, though.

**Comment by Andreya-Autumn** [comment]
Right, so it's not aliasing it's zippering. Let's just fix that yeah? Strictly it's a breaking change but I doubt anyone will mind.

**Comment by mkruselj** [comment]
Yep I'll open the PRs.

**Comment by mkruselj** [comment]
`sst-basic-blocks` PR: https://github.com/surge-synthesizer/sst-basic-blocks/pull/308

- [x] Need that one reviewed and merged before I open a PR towards Surge.
